File: backend-spring/src/main/java/com/ivyts/backend/service/LearningService.java
package com.ivyts.backend.service;

import com.ivyts.backend.common.exception.ApiException;
import com.ivyts.backend.domain.course.Course;
import com.ivyts.backend.domain.enrollment.Enrollment;
import com.ivyts.backend.domain.lesson.Lesson;
import com.ivyts.backend.domain.user.User;
import com.ivyts.backend.domain.user.UserRole;
import com.ivyts.backend.security.AuthUser;
import com.ivyts.backend.service.coursestore.CourseStore;
import com.ivyts.backend.service.coursestore.LessonStore;
import com.ivyts.backend.service.enrollmentstore.EnrollmentStore;
import com.ivyts.backend.service.userstore.UserStore;
import com.ivyts.backend.web.course.CourseMapper;
import com.ivyts.backend.web.enrollment.EnrollmentMapper;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

@Service
public class LearningService {

    private final CourseStore courseStore;
    private final LessonStore lessonStore;
    private final EnrollmentStore enrollmentStore;
    private final UserStore userStore;
    private final CourseMapper courseMapper;
    private final EnrollmentMapper enrollmentMapper;

    public LearningService(
        CourseStore courseStore,
        LessonStore lessonStore,
        EnrollmentStore enrollmentStore,
        UserStore userStore,
        CourseMapper courseMapper,
        EnrollmentMapper enrollmentMapper
    ) {
        this.courseStore = courseStore;
        this.lessonStore = lessonStore;
        this.enrollmentStore = enrollmentStore;
        this.userStore = userStore;
        this.courseMapper = courseMapper;
        this.enrollmentMapper = enrollmentMapper;
    }

    public Map<String, Object> getLearningData(String courseId, AuthUser authUser) {
        if (authUser.role() != UserRole.STUDENT) {
            throw new ApiException(HttpStatus.FORBIDDEN, "Only students can access learning data");
        }

        Course course = courseStore.findById(courseId)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Course not found"));
        Enrollment enrollment = enrollmentStore.findByCourseAndStudent(courseId, authUser.userId())
            .orElseThrow(() -> new ApiException(HttpStatus.FORBIDDEN, "You must enroll in this course before accessing learning data"));

        List<Lesson> lessons = lessonStore.findByCourseOrderByOrderAsc(courseId);
        Lesson currentLesson = lessons.stream()
            .filter(lesson -> lesson.getId().equals(enrollment.getLastLessonId()))
            .findFirst()
            .orElse(lessons.isEmpty() ? null : lessons.get(0));
        Lesson nextLesson = currentLesson == null
            ? null
            : lessons.stream().filter(lesson -> lesson.getOrder() == currentLesson.getOrder() + 1).findFirst().orElse(null);

        User owner = findUserOrThrow(course.getOwner());
        User student = findUserOrThrow(authUser.userId());

        Map<String, Object> data = new LinkedHashMap<>();
        data.put("course", courseMapper.toCourseView(course, owner, lessons));
        data.put("lessons", lessons.stream().map(courseMapper::toLessonView).toList());
        data.put("enrollment", enrollmentMapper.toEnrollmentView(enrollment, student, course));
        data.put("currentLessonId", currentLesson != null ? currentLesson.getId() : null);
        data.put("nextLessonId", nextLesson != null ? nextLesson.getId() : null);
        return data;
    }

    private User findUserOrThrow(String userId) {
        return userStore.findById(userId)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "User not found"));
    }
}

File: archive/backend-node/src/services/learning.service.ts
// Builds the one-shot payload used by the full-screen learning page after verifying enrollment access.
import { Course, Enrollment, Lesson } from '../models/index.js';
import { AuthUser } from '../types/auth.js';
import { HTTP_STATUS } from '../constants/http-status.js';
import { ApiError } from '../utils/api-error.js';
import { mapCourse, mapLesson } from '../utils/course-transform.js';
import { mapEnrollment } from '../utils/enrollment-transform.js';

export async function getLearningData(courseId: string, user: AuthUser) {
  const course = await Course.findById(courseId).populate('owner', 'fullName email');

  if (!course) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Course not found');
  }

  const enrollment = await Enrollment.findOne({
    course: courseId,
    student: user.userId,
  })
    .populate('course', 'title slug thumbnail category level lessonCount totalDuration')
    .populate('student', 'fullName email');

  if (!enrollment) {
    throw new ApiError(HTTP_STATUS.FORBIDDEN, 'You must enroll in this course before accessing learning data');
  }

  const lessons = await Lesson.find({ course: course._id }).sort({ order: 1 });
  const currentLesson =
    lessons.find((lesson) => lesson._id.toString() === enrollment.lastLessonId?.toString()) ?? lessons[0] ?? null;
  const nextLesson =
    currentLesson
      ? lessons.find((lesson) => lesson.order === currentLesson.order + 1) ?? null
      : null;

  return {
    course: mapCourse(course, lessons),
    lessons: lessons.map((lesson) => mapLesson(lesson)),
    enrollment: mapEnrollment(enrollment),
    currentLessonId: currentLesson?._id.toString(),
    nextLessonId: nextLesson?._id.toString(),
  };
}
